src/event/epoll.rs

```diff
@@ -0,0 +1,495 @@
+#[inline]
+#[doc(alias = "epoll_create1")]
+pub fn create(flags: CreateFlags) -> io::Result<OwnedFd> {
+    syscalls::epoll_create(flags)
+}
+
+/// `epoll_ctl(self, EPOLL_CTL_ADD, data, event)`—Adds an element to an epoll
+/// object.
+///
+/// This registers interest in any of the events set in `events` occurring on
+/// the file descriptor associated with `data`.
+///
+/// Note that `close`ing a file descriptor does not necessarily unregister interest
+/// which can lead to spurious events being returned from [`wait`]. If a file descriptor
+/// is an `Arc<dyn SystemResource>`, then `epoll` can be thought to maintain a
+/// `Weak<dyn SystemResource>` to the file descriptor. Check the [faq] for details.
+///
+/// # References
+///
+///  - [Linux]
+///
+/// [Linux]: https://man7.org/linux/man-pages/man2/epoll_ctl.2.html
+/// [faq]: https://man7.org/linux/man-pages/man7/epoll.7.html#:~:text=Will%20closing%20a%20file%20descriptor%20cause%20it%20to%20be%20removed%20from%20all%0A%20%20%20%20%20%20%20%20%20%20epoll%20interest%20lists%3F
+#[doc(alias = "epoll_ctl")]
+#[inline]
+pub fn add(
+    epoll: impl AsFd,
+    source: impl AsFd,
+    data: EventData,
+    event_flags: EventFlags,
+) -> io::Result<()> {
+    syscalls::epoll_add(
+        epoll.as_fd(),
+        source.as_fd(),
+        &mut Event {
+            flags: event_flags,
+            data,
+            #[cfg(all(libc, target_os = "redox"))]
+            _pad: 0,
+        },
+    )
+}
+
+/// `epoll_ctl(self, EPOLL_CTL_MOD, target, event)`—Modifies an element in a
+/// given epoll object.
+///
+/// This sets the events of interest with `target` to `events`.
+///
+/// # References
+///
+///  - [Linux]
+///
+/// [Linux]: https://man7.org/linux/man-pages/man2/epoll_ctl.2.html
+#[doc(alias = "epoll_ctl")]
+#[inline]
+pub fn modify(
+    epoll: impl AsFd,
+    source: impl AsFd,
+    data: EventData,
+    event_flags: EventFlags,
+) -> io::Result<()> {
+    syscalls::epoll_mod(
+        epoll.as_fd(),
+        source.as_fd(),
+        &mut Event {
+            flags: event_flags,
+            data,
+            #[cfg(all(libc, target_os = "redox"))]
+            _pad: 0,
+        },
+    )
+}
+
+/// `epoll_ctl(self, EPOLL_CTL_DEL, target, NULL)`—Removes an element in a
+/// given epoll object.
+///
+/// # References
+///
+///  - [Linux]
+///
+/// [Linux]: https://man7.org/linux/man-pages/man2/epoll_ctl.2.html
+#[doc(alias = "epoll_ctl")]
+#[inline]
+pub fn delete(epoll: impl AsFd, source: impl AsFd) -> io::Result<()> {
+    syscalls::epoll_del(epoll.as_fd(), source.as_fd())
+}
+
+/// `epoll_wait(self, events, timeout)`—Waits for registered events of
+/// interest.
+///
+/// For each event of interest, an element is written to `events`. On
+/// success, this returns the number of written elements.
+///
+/// # References
+///
+///  - [Linux]
+///
+/// [Linux]: https://man7.org/linux/man-pages/man2/epoll_wait.2.html
+#[inline]
+pub fn wait<B: buf::EventBuffer>(
```

Comment:
I would keep the `epoll_` prefix here. It's a bit weird to have freestanding functions just named "wait".

Reply:
I think the idea is to call it with the module prefix like `epoll::wait` (implemented here: https://github.com/bytecodealliance/rustix/commit/ac4a66f05f2976b21a603135da31eb9d04f6b2a5). But anyway, can't change that as it would be a breaking change.

Reply:
Hmmm, looking at the rest of our APIs I think yours is more consistent. Filed https://github.com/bytecodealliance/rustix/issues/753#issuecomment-2292528359